### DISCUSSION TITLE
not sure how to fix this - Debrick

### DISCUSSION THREAD

**Discussion by gradientledge** [bangcorrupt/hacktribe#79] [General]
so updated the firmware and on reboot i get the bootscreen hacktribe with a glitch in the H of the hacktribe then asks for software update is there anyway to fix this, because i've made the same SD card file order but says error file not found, do i have to crack the case and jump it or can i somehow reflash this.

**Comment by bangcorrupt** [comment]
Assuming you had a synth originally (grey or blue), you should only need the synth firmware under the synth directory.  See [Boots to prompt](https://github.com/bangcorrupt/hacktribe/wiki/Debrick#boots-to-prompt) in the wiki.  

It's odd that it has the hacktribe logo but still gets to the device ID check, how did you create the firmware that caused this?

**Comment by gradientledge** [comment]
i must of somehow took an old image thats all i can imagine and did the
powershell patch maybe i used an already hacked renamed image but it
wouldn't have went through if that was the case i have no idea or my sd
card didnt read everything on the update but that shouldn't of happened.

**Reply by gradientledge** [reply]
im pointing to my sd now because everytime i tried to update with stock fw it would go through the working then say no card so its probably safe to say this reader is toast, ill pick up a new card today and try it, its just FAT32 with Korg/Electribe/Synth/System correct just because theres no way i can format the new card from the device in current state.

**Reply by bangcorrupt** [reply]
> Korg/Electribe/Synth/System

`KORG/electribe/System/SYSTEM.VSB`

KORG is all caps, electribe is all lowercase, no Synth directory.   

If you put factory synth firmware version 2.02  at that path it should just work, if it doesn't I have no idea why.  If you email me the update file that failed I can install it and find a fix.

**Reply by gradientledge** [reply]
ok ill try to find it, last night was like a hectic try to find a fix lol, but then odd thing was knob turns still registered while software update prompt was up which seemed like ok something installed partially, not a full install ill send it over.

**Reply by gradientledge** [reply]
so put everything as with the original synth file structure with the 2.02 and it says no file even with new card so i have no idea maybe its looking for the sampler version

**Reply by bangcorrupt** [reply]
It could be, if it somehow performed half an update without completely breaking.  If you zip up the KORG directory and email it to me along with the broken firmware I can take a look.

**Reply by bangcorrupt** [reply]
The firmware you installed is shuffled and mixed up like crazy. I'm not sure what happened to create this, it's interesting that it boots as far as it does.  When it checks for a valid update file it uses a device ID that shouldn't exist (0x126), I can't see a way of changing this without a debugger attached.  Long story short, this can be fixed by following the full [debrick guide](https://github.com/bangcorrupt/hacktribe/wiki/Debrick#rpi-and-openocd).  There will be a more accessible guide at some point, probably in a few weeks.

**Comment by gradientledge** [comment]
ok i appreciate it im just glad there's a fix ill try to source up a jtag
connector and make sure i have everything to do it.

**Reply by bangcorrupt** [reply]
> jtag connector

A raspberry pi is ideal, especially if you don't have a spare electribe to dump the bootloader from.  The JTAG header is 6-pin, 1.5mm pitch and you'll need the crimp tool if you want to make a proper lead.  The first time I did it I just poked some 22 AWG wire in the holes and hoped for the best...

**Comment by gradientledge** [comment]
ok thanks i can source a raspbi pi i have plenty of arduino nanos and
components around here i take it you need a bigger pie for the header, will
a zero w work if you have a breadboard to use for more pins

**Reply by bangcorrupt** [reply]
I use a pi zero W via a breadboard.   If you solder a [1.5mm JST ZH header](https://uk.rs-online.com/web/p/pcb-headers/1667270) to the e2, you will need to step up to the breadboard 2.54mm pitch somehow.  When I soldered mine, the contacts on the board were oxidised and it ended up a mess.  It may be worth cleaning with some wire wool or something first.

**Reply by gradientledge** [reply]
ok sweet i work as smt operator so thats no problem to source or even further down the line create a solid debugger for this that can be easy for anyone else. appreciate the help, i might catalog pics and ways to do this with bare minimum parts for your tutorial, to make the hardware portion less intimidating.

**Reply by bangcorrupt** [reply]
> catalog pics

Please do this :smile:.  Answering these questions makes it easier to work out what needs to be in the tutorial.  Tips on clean soldering would be helpful too, I'm far from pro.

It will also work with a cheap JLink clone, but requires a backup of the bootloader from a working device first.  When you attach a debugger before powering up it hangs in the ROM Bootloader in ARM RAM.  The JLink is attached as soon it is powered.  With openocd on rpi, it only attaches when you run the command, so if we're quick we can halt in the Secondary Bootloader in On-Chip RAM, before handing off to the corrupt app.

**Reply by gradientledge** [reply]
ok sounds good ill email pics over, im far from pro as well but i can kinda get how this is going, im sure after you do it once its probably cake i cant imagine the number of times you attempted this to get it to work, so once you pause the boot from the rpi do you then direct it to load the correct firmware im trying to piece this together, as to how the fix works in a nutshell.

**Reply by bangcorrupt** [reply]
Get it to halt in the bootloader by ~running the `openocd` command at the same time as pressing the power button.~ following the steps in the [debrick guide](https://github.com/bangcorrupt/hacktribe/wiki/Debrick#get-secondary-bootloader-sbl). Then, connect to the telnet server provided by openocd: `telnet localhost 4444`, set a breakpoint at 0xc0000000: `bp 0xc0000000 4` and continue: `resume`.  (I can probably automate setting the breakpoint in the config file).   Execution should halt at the breakpoint at 0xc0000000.

At 0xc0000000 (external DDR RAM), the bootloader has finished loading the app from the flash and the first instruction of the app is about to be executed.  We halt here and overwrite the broken app in RAM with the data from the firmware update file (minus header): `load_image SYSTEM.bin 0xc0000000`.  (There's probably a way to specify a file offset instead of chopping header off the file before hand).  

Then when we continue, the app we just loaded into RAM should boot correctly.  Once booted, we can use the normal firmware update function to write the flash.

**Reply by gradientledge** [reply]
so i read the openocd article but im confused of the pins locations on the pi zero 2w and where they connect on the electribe pcb j9, do i run the wires directly from the pi to j9 or do i have to have the resistors and clock on breadboard as well? appreciate it.

**Reply by bangcorrupt** [reply]
We can connect the rpi gpio directly to the e2 j9 header, they are both 3.3V.  That article is good for understanding the config files, most of the hardware and the SWD stuff is not necessary.  

I pushed a change to [openocd.cfg](https://github.com/bangcorrupt/hacktribe/blob/dev/openocd/openocd.cfg), the reset pin was specified incorrectly.  In that file, the comment at line 20 `# Header pin numbers: 23 22 19 21` shows the physical location of the [pins on the header](https://pinout.xyz/), the command at line 21 `bcm2835gpio_jtag_nums 11 25 10 9` uses the pin numbers as the CPU sees them.  

Using pin numbers by physical location:

    E2.J9 <-> RPI.GPIO

    GND <---> 20 (or any other ground pin)
    TMS <---> 22
    TCK <---> 23
    TRS <---> 26
    TDI <---> 19
    TDO <---> 21

I think we can remap any pin (except ground) as long as it is specified in the config file.

**Reply by gradientledge** [reply]
nice that makes way more sense, thanks.  I still have to get a raspberry, but sounds way less intimidating now.

**Comment by bangcorrupt** [comment]
I realised there is an important step missing from the wiki, we need a toggle switch to keep the power supply active before the firmware boots. See https://github.com/bangcorrupt/hacktribe/discussions/216#discussioncomment-4632545 for some more details.
